modules/capability-log-write/apis/composition.yaml

```diff
@@ -0,0 +1,188 @@
+            kind: Object
+            metadata:
+              annotations:
+                gotemplating.fn.crossplane.io/composition-resource-name: remote-environment-config-consumers-{{ $cluster }}
+                gotemplating.fn.crossplane.io/ready: "True"
+            spec:
+              providerConfigRef:
+                name: kubernetes-{{ $cluster }}
+              managementPolicies: [Observe]
+              watch: true
+              forProvider:
+                manifest:
+                  apiVersion: apiextensions.crossplane.io/v1beta1
+                  kind: EnvironmentConfig
+                  metadata:
+                    name: consumers.logwrites.capabilities.elastisys.com
+            {{- end }}
+    - step: patch-consumer-secrets
+      functionRef:
+        name: function-go-templating
+      input:
+        apiVersion: gotemplating.fn.crossplane.io/v1beta1
+        kind: GoTemplate
+        source: Inline
+        inline:
+          template: |
+            {{- $isManager := .observed.composite.resource.spec.clusters }}
+            {{- if $isManager }}
+            {{- $producer := "" }}
+            {{- $producers := index (index .context "apiextensions.crossplane.io/extra-resources") "producers" }}
+            {{- if and $producers (eq (len $producers) 1) }}
+            {{- $producer = index $producers 0 }}
+            {{- end }}
+            {{- $clusters := list }}
+            {{- $localConsumers := index (index .context "apiextensions.crossplane.io/extra-resources") "consumers" }}
+            {{- if $localConsumers }}
+              {{- $cluster := dict
+                "providerConfigRef" "kubernetes"
+                "secrets" list
+              }}
+              {{- range $_, $secret := $localConsumers }}
+                  {{- $cluster = set $cluster "secrets" (append $cluster.secrets (dict "name" $secret.metadata.name "namespace" $secret.metadata.namespace)) }}
+              {{- end }}
+              {{- $clusters = append $clusters $cluster }}
+            {{- end }}
+            {{- if hasKey .observed "resources" }}
+            {{- range $_, $remoteClusterName := .observed.composite.resource.spec.clusters }}
+            {{- with index $.observed.resources (print "remote-environment-config-consumers-" $remoteClusterName) }}
+            {{- $remoteConsumers := .resource.status.atProvider.manifest.data.secretRefs }}
+            {{- if $remoteConsumers }}
+              {{- $cluster := dict
+                "providerConfigRef" (print "kubernetes-" $remoteClusterName)
+                "secrets" list
+              }}
+              {{- range $_, $secret := $remoteConsumers }}
+                  {{- $cluster = set $cluster "secrets" (append $cluster.secrets (dict "name" $secret.name "namespace" $secret.namespace)) }}
+              {{- end }}
+              {{- $clusters = append $clusters $cluster }}
+            {{- end }}
+            {{- end }}
+            {{- end }}
+            {{- end }}
```

Comment:
This is a lot of logic that, in a proper implementation, should be implemented in a function rather than in Go templating. :smile: Also worth considering if Module Capabilities might require enough functionality to warrant it's own Crossplane Provider / Kubernetes controller.